Mathlib/Data/ListM/Basic.lean

```diff
@@ -361,6 +361,82 @@ This will run forever if the list is infinite. -/
 unsafe def fold (f : β → α → β) (init : β) (L : ListM m α) : m β :=
   L.foldM (fun b a => pure (f b a)) init
 
+/--
+Extract the prefix up a lazy list consisting of elements up to and including
+the first element satisfying a monadic predicate.
+Return (in the monad) the prefix as a `List`, along with the remaining elements as a `ListM`.
+-/
+partial def getUpToFirstM (L : ListM m α) (p : α → m (ULift Bool)) : m (List α × ListM m α) := do
+  match ← L.uncons with
+  | none => return ([], nil)
+  | some (x, xs) => (if (← p x).down then
+      return ([x], xs)
+    else do
+      let (acc, R) ← getUpToFirstM xs p
+      return (x :: acc, R))
+
+/--
+Extract the prefix up a lazy list consisting of elements up to and including
+the first element satisfying a predicate.
+Return (in the monad) the prefix as a `List`, along with the remaining elements as a `ListM`.
+-/
+def getUpToFirst (L : ListM m α) (p : α → Bool) : m (List α × ListM m α) :=
+  L.getUpToFirstM fun a => pure (.up (p a))
+
+/--
+Extract a maximal prefix of a lazy list consisting of elements
+satisfying a monadic predicate.
+Return (in the monad) the prefix as a `List`, along with the remaining elements as a `ListM`.
+
+(Note that the first element *not* satisfying the predicate will be generated,
+and pushed back on to the remaining lazy list.)
+-/
+partial def getWhileM (L : ListM m α) (p : α → m (ULift Bool)) :
```

Comment:
Should this be `takeWhile`, or is the divergence from the list name deliberate?

Reply:
I think this function is to `takeWhile` as `splitAt` is to `take` (it returns both parts and not just the rest).  So maybe `splitAfter`?

Reply:
How about `splitWhile`? I would like `while` in the name.